### PR TITLE
Alerting: Special alert instance if rule is in state NoData

### LINKED ISSUE #36376: Ngalert in Grafana 8 is missing option to "Keep Last State"
**What happened**:

The new alerting introduced with Grafana 8 is missing the "Keep Last State" option:
alerting_rules

No Data Option | Description
-- | --
No Data | Set alert state to NoData and rule state to Normal
Alerting | Set alert rule state to Alerting
Ok | Set alert rule state to Normal

**What you expected to happen**:
The legacy alerting system had the following options:
No Data Option | Description
-- | --
No Data | Set alert rule state to NoData
Alerting | Set alert rule state to Alerting
**Keep Last State** | **Keep the current alert rule state, whatever it is.**
Ok | Not sure why you would want to send yourself an alert when things are okay, but you could.

**Environment**:
- Grafana version: v8.0.3 (cae5c5e46b)
- OS Grafana is installed on: linux
- User OS & Browser: windows, chrome

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**Background**:
The legacy version of Grafana Alerts supports several ways of handling how we deal with no data (or execution errors) as we query a data source for alerting data. When the data source returns no data or null values we have:
- `Alerting` -  set the state to Alerting, and therefore, notify the subscribers. No notification will be sent if the alert is already in this state. Otherwise, a generic alert with no metrics in the body will be sent.
- `No Data` - set the state to “no data”. This will trigger a notification if the alert is in OK or Alerting state.
- `Keep Last State` - The alerting engine keeps the last state of the alert. Even if it is in a firing state.
- `OK` - change state to OK, ie if it is alerting it will send notification.

Unified alerting removes support for “Keep Last State”, and during migration from the legacy alerting engine, alerts that are configured with the option `Keep Last State` is changed to `Alerting`. This is a decision that the unified alerting took deliberately to simplify the transition between systems, at the time the team didn’t understand the use-cases for this state.

However, during the beta period, some users came forward on the usage of this option. One use-case is to reduce the possibility of a misfire when data points are missing in the observable interval. Another is that users leverage this option by creating two alert rules: one with “Keep Last state” that is sent to whoever is interested in the parameter that triggers the alert, and another with “Alerting” that notifies the operations team who would deal with data source availability.

During the beta period, some users came forward on the usage of this option. One use-case is to reduce the possibility of a misfire when data points are missing in the observable interval (https://github.com/grafana/grafana/issues/36376#issuecomment-874488534). Another is that users leverage this option by creating two alert rules: one with “Keep Last state” that is sent to whoever is interested in the parameter that triggers the alert, and another with “Alerting” that notifies the operations team who would deal with data source availability (https://github.com/grafana/grafana/issues/39056#issuecomment-917119707).

The migration to Unified Alerting can cause alert rules that use “Keep Last State” to start firing right after migration due to the option not being available anymore, which can trigger unexpected incidents. Some users have already expressed their concerns about the removal of the option firing alerts that weren’t firing before.
Given that Unified Alerting decouples alert evaluation and state management from alert routing and grouping. It becomes a two-part problem to support something similar to the “Keep Last State” behavior.
The added feature of routing and grouping is now supported by a label-based system often referred to as “multi-dimensional” alerting. This is important as alerts are no longer notified 1-1, but instead routed through a tree-like structure based on the label set. Where each set counts as 1 alert and an alert rule can output multiple sets.

After [discussion ](https://docs.google.com/document/d/1k5Bo-yn-EY_jLi08QFzt28VEK1KquQAkTR78lbMmw60/edit#) in the alerting team, we decided that the better approach would be to 
1. introduce a fixed label-set (Alert) when an alert rule goes into “no data”. This would replace the actual label set of this alert. 
2. during migration from the legacy alerting system, create a silence that would match the label set of the new dedicated alert.

**What this PR does / why we need it**:
This PR updates unified alerting engine:
- the state manager to allow state NoData to be sent to notifications.
- conversion of an alert state to alert instance (a model that is submitted to Alertmanager) to generate a special alert when the alert state's evaluation state is NoData:
   - copy the original name of the alert to label  `original_alertname`
   - update label `alertname` to constant value `DatasourceNoData`

Updates migration of alert that has the option "Keep Last State" for evaluation state `NoData`
- to change to option "No Data" (current is "Alerting")
- adds a silence that matches the alert by labels `alertName` and `rule_uid`.

<details><summary>Screentshots</summary>
Page "Alert rules":

image

Page "Alert Groups":
image

Page "Silences"
image

</details>

**Which issue(s) this PR fixes:**
Fixes: https://github.com/grafana/grafana/issues/36376
Fixes: grafana/alerting-squad#178
Alternative approach: https://github.com/grafana/grafana/pull/40494.

**Special notes for your reviewer**:
There are a few known issues that need to be addressed:
1. Firing alert (whether it is regular Alerting or NoData) is not deleted right after the state is changed because the scheduler does not expire existing alerts. Instead, it just pushes a new alert instance and relies on the fact that the instance will overwrite the existing one. However, in this case, it will not work because two alerts (created from Alerting and NoData states) have a different set of labels. Therefore, there is a high chance of two alert instances firing at the same time. This is another edge case of https://github.com/grafana/grafana/issues/39154 which will be addressed as follow up after https://github.com/grafana/grafana/pull/39975 is merged
2. UI does not display the NoData alert instance in the page "Alert Rules". 
3. UI displays status NoData as a "ok" status, but this PR makes it to be the same severity as Alerting. We need to change color of the status.

### CODE REVIEW

File: pkg/services/ngalert/state/state.go
Comment (gerobinson) [inline pkg/services/ngalert/state/state.go:117]: I think I understand the intention of this change but I just wanted to check that what we are doing is changing `a.State != eval.Alerting && a.State != eval.Normal` to `a.State == eval.Pending || a.State == eval.Error` such that `eval.NoData` will return true (depending on `LastEvaluationTime`).

File: pkg/services/ngalert/schedule/compat.go
Comment (gerobinson) [inline pkg/services/ngalert/schedule/compat.go:17]: Extra newline here?

File: pkg/services/sqlstore/migrations/ualert/silences.go
Comment (gerobinson) [inline pkg/services/sqlstore/migrations/ualert/silences.go:59]: If a user later changes the no data state for this alert from `Keep Last State` to something like `Alerting` do we expire the old silence?

Comment (gerobinson) [inline pkg/services/sqlstore/migrations/ualert/silences.go:59]: If a user later changes the no data state for this alert from `Keep Last State` to something like `Alerting` do we expire the old silence?

**Review by grobinson-grafana — APPROVED** [review]
Looks good, I have just a couple of questions. I think we should wait on a review from @gotjosh too.

File: pkg/services/sqlstore/migrations/ualert/silences.go
Comment (yuri-tceretian) [reply pkg/services/sqlstore/migrations/ualert/silences.go:59]: Good question! Currently, we do not expire the silence when the rule is changed. However, the silence will not match any alert instances. So, it will become kinda useless. I am hesitant whether we should do that because ngalert does not have that relationship between silence and alert rule: users have to create a rule and then a silence for NoData state separately.

Comment (gerobinson) [reply pkg/services/sqlstore/migrations/ualert/silences.go:59]: Won't the silence match the `DatasourceNoData` alert? I might have misunderstood but I thought we wanted notifications for that alert when we change no data state from `Keep Last State` to something else?

File: pkg/services/ngalert/state/state.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/state/state.go:117]: Correct, the previous logic always returned false to NoData. I merged two conditions and rewrote with equivalent disjunction which is a little bit more readable as it does not use negations. Then I just removed NoData from the expression.

**Review by grobinson-grafana — APPROVED** [review]

**Comment by gotjosh** [comment]
I'm not saying the approach here is wrong and I need a bit more time to test it, but in a nutshell, I'm not sure I get why this work is being done right before sending an alert (by modifying the original alert) as opposed to the state manager (by emitting a completely new alert that is attached to the datasource(s) "status").

I'm not saying this is wrong, but I need to understand it a bit better.

I have two arguments or trains of thought that I would love clarification on (I'll do some more testing tomorrow but going to leave this as a question for now).

1. Can we deterministically know what are labels this alert is going to have in the alertmanager? Because we're using the same alert that would be sent and then replacing 1 of its labels, we're at the mercy of whatever results the query produced. I believe that not all data sources produce a consistent output when we have "nodata" and there's a non-zero chance that labels and annotation expansion did not happen. To top that off, I'd hope that this work would include the information about the failing datasource. 

2. It is not clear to me what happens **to both** the existing alert and new alert **on both** the scheduler/state manager and the alertmanager, can you please break down or create some tests for it? For example, here are a few cases:

Let’s say that I have an alert every 10s for 60s
- case 1 (no data between internals)
    - t+10s -> ok, t+10s -> ok, t+10s -> no-data, t+10s -> no-data, t+10s -> ok, t+10s -> ok
- case 2 (no data after firing)
    - t+50s -> pending, t+60s -> alerting, t+70s -> no-data, etc
- case 3 (no data just before firing)
    - t+50s -> pending, t+60s -> no-data, t+70s -> ???

How does the above change for an alert that's 1m every 5m?

**Comment by yuri-tceretian** [comment]
I will address your comment in more detail tomorrow but they seem to be related to the state management. The reason I changed the conversion logic is to not touch the state management because it could cause more fundamental changes. 
To answer your question about how we know what labels are going to be in the case of NoData: if the evaluation engine returns a result that has the same labels but no data point, and assuming that there are no labels that mention the value in the template, the resulting labels will be the same as regular Alerting state with exception of two labels: `alertName` and `original_alert_uid`. When the evaluation result returns the different set of labels, the state manager will create a separate group for it, which won't affect the other states of this alert. 
Also, I believe that the alert instance Grafana will generate in the case of NoDataState = NoData is pretty much the same as when NoDataState = Alerting (again, except two labels: `alertName` and `original_alert_uid`).

**Comment by yuri-tceretian** [comment]
> Let’s say that I have an alert every 10s for 60s
>case 1 (no data between internals)
>t+10s -> ok, t+10s -> ok, t+10s -> no-data, t+10s -> no-data, t+10s -> ok, t+10s -> ok
>case 2 (no data after firing)
>t+50s -> pending, t+60s -> alerting, t+70s -> no-data, etc
>case 3 (no data just before firing)
>t+50s -> pending, t+60s -> no-data, t+70s -> ???
>How does the above change for an alert that's 1m every 5m?


In the case if the evaluation result provides a label (or many), all cases above would result in state reset when NoData happens because the CacheId changes. I implemented tests to emphasize this behavior https://github.com/grafana/grafana/pull/41291

**Comment by yuri-tceretian** [comment]
The diagrams below are based on assumption that the NoData evaluation result has the same labels as the state (ie CacheId matches the previous evaluations). Also, let's assume the rule interval is every 10 seconds for 20 seconds.

> case 3 (no data just before firing)
> t+50s -> pending, t+60s -> no-data, t+70s -> ???

 This is what happens (see [test])(https://github.com/grafana/grafana/pull/41291/commits/2eb6efd2003b0aed975a136eb39ade6ba5f0f72f)
image

The expected behavior of Grafana in this PR is that at T+10 the it will start firing alert instance `DatasourceNoData`. The fact that the evaluation result had the same labels as the previous evaluation means that all templates will be expanded as they would be in the case of alerting. The `DatasourceNoData` alert will stop firing in 90 seconds because the current implementation of alerting will not expire it immediately.


>case 2 (no data after firing)
>t+50s -> pending, t+60s -> alerting, t+70s -> no-data, etc

The diagram would look like (see [test](https://github.com/grafana/grafana/pull/41291/commits/97213a17c2630f23bdadd26200b2c1e4db603d5e)
image


>case 1 (no data between internals)
>t+10s -> ok, t+10s -> ok, t+10s -> no-data, t+10s -> no-data, t+10s -> ok, t+10s -> ok

image

**Comment by yuri-tceretian** [comment]
>  To top that off, I'd hope that this work would include the information about the failing datasource.

Initially, I implemented it but then decided to remove it because I think it will require changes in API and I wanted this PR to be merged to confirm the approach. The problem with `Error state` is that there is no good option for sending `DatasourceError` alert. Currently, there are two options:
- `OK` - which basically allows users to ignore the errors
- `Alerting` - which basically converts error to a regular Alerting state which will affect Pending state.
I think we need another option `Error` which meaning will be similar to `NoData` that is introduced in this PR.

File: pkg/services/sqlstore/migrations/ualert/silences.go
Comment (gotjosh) [inline pkg/services/sqlstore/migrations/ualert/silences.go:87]: We're using the title here but the UID above, should we use both?

```suggestion
			Comment:   fmt.Sprintf("Created during migration to unified alerting to silence NoData state for alert rule ID '%s' and Title '%s' because the option 'Keep Last State' was selected for NoData state", rule.UID, rule.Title),
```

**Review by gotjosh — APPROVED** [review]
LGTM - all of my comments are nits. Let me know if you want me to take a second look but I don't have to.

File: pkg/services/ngalert/schedule/compat.go
Comment (gotjosh) [inline pkg/services/ngalert/schedule/compat.go:56]:
```suggestion
	if alertState.State == eval.NoData {
		return noDataAlert(nL, nA, alertState, urlStr)
	}
	
// NoDataAlert is a special alert sent by Grafana to the Alertmanager, that indicates we received no data from the datasource.
// It effectively replaces the legacy behavior of "Keep Last State" by separating the regular alerting flow from the no data scenario into a separate alerts.
// The Alert is defined as:
// {  alertname=DatasourceNoData rulename=original_alertname } + { rule labelset } + { rule annotations }
func noDataAlert(labels data.Labels, annotations data.Labels, alertState *state.State, urlStr string) *models.PostableAlert {
	if name, ok := labels[model.AlertNameLabel]; ok {
		labels[OriginalAlertName] = name
	}
	labels[model.AlertNameLabel] = NoDataAlertName

	return &models.PostableAlert{
		Annotations: models.LabelSet(annotations),
		StartsAt:    strfmt.DateTime(alertState.StartsAt),
		EndsAt:      strfmt.DateTime(alertState.EndsAt),
		Alert: models.Alert{
			Labels:       models.LabelSet(labels),
			GeneratorURL: strfmt.URI(urlStr),
		},
	}
}
```

Comment (gotjosh) [reply pkg/services/ngalert/schedule/compat.go:56]: What do you think about this? The only important part of my suggestion is the comment - I would like to make it crystal clear on how this part works.

**Comment by gotjosh** [comment]
Sorry for sending you on a wild goose chase @yuri-tceretian with my last comment. It seems I misunderstood a part of how this works and thought it would work a bit differently.

I'm perfectly OK with this approach, good work 👏

File: pkg/services/sqlstore/migrations/ualert/silences.go
Comment (yuri-tceretian) [reply pkg/services/sqlstore/migrations/ualert/silences.go:85]: we already use UID in the silences (a function above the one I added)

>  I'm not sure if ID's will stay the same across the databases

across databases? do you mean tables (legacy alert and ngalert)?

Comment (gotjosh) [reply pkg/services/sqlstore/migrations/ualert/silences.go:85]: Different databases - mySQL, postgres, etc. Unsure which field we use and how this translates to the ORM. I assume we simply use a `VARCHAR` fields but you never know 😅

Comment (gotjosh) [reply pkg/services/sqlstore/migrations/ualert/silences.go:85]: If we have precedent (we use it in the above function) then I've missed it and this is completely fine.

**Comment by gotjosh** [comment]
This failed due to https://github.com/grafana/grafana/pull/39855 mofiying `compat.go` and not being backported.

I have backported the previous PR and will now proceed to backport this one.

**Pull request merged**
